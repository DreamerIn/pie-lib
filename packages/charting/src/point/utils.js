import debug from 'debug';
import omitBy from 'lodash/omitBy';
import isUndefined from 'lodash/isUndefined';

const log = debug('pie-lib:charting:point:utils');

/**
 * Get the index of the point in the points array
 * @param {Point[]} points
 * @param {Point} p
 */
export const pointIndex = (points, p) => {
  return (points || []).findIndex(s => {
    return s.x === p.x && s.y === p.y;
  });
};

export const hasPoint = (points, p) => pointIndex(points, p) !== -1;

export const trim = point => ({
  x: point.x,
  y: point.y,
  label: point.label
});

export const removePoints = (points, toRemove) => {
  log('[removePoints] points: ', points, toRemove);
  toRemove = Array.isArray(toRemove) ? toRemove : [toRemove];

  const out = points.filter(p => !hasPoint(toRemove, p));

  log('[removePoints] points: ', out);
  return out;
};

export const firstAvailableLabel = (points, labels) => {
  const existing = points.map(p => p.label);
  log('points: ', points);
  const filtered = labels.filter(l => !existing.some(e => e === l));
  log('existing: ', existing, 'filtered: ', filtered);
  return filtered.length > 0 ? filtered[0] : '';
};

export const addPoint = (points, p, labels) => {
  labels = labels || [];
  if (!hasPoint(points, p)) {
    const out = points.slice();
    const label = firstAvailableLabel(points, labels);
    out.push(Object.assign({}, p, { label }));
    return out;
  } else {
    return points;
  }
};

export const bounds = (p, domain, range) => {
  return {
    left: domain.min - p.x,
    right: Math.abs(p.x - domain.max),
    top: Math.abs(p.y - range.max),
    bottom: range.min - p.y
  };
};

export const swap = (arr, from, to, getIndex) => {
  log('[swap] arr: ', arr, 'from: ', from, 'to: ', to);

  const toIndex = getIndex(arr, to);

  log('[swap] toIndex: ', toIndex);

  if (toIndex !== -1) {
    //prevent moving one point on top of another.
    return arr;
  }

  const fromIndex = getIndex(arr, from);

  log('[swap] fromIndex: ', fromIndex);

  if (fromIndex === -1) {
    //cant move a point that isn't in the points array
    return arr;
  }

  const f = arr[fromIndex];
  const replacement = Object.assign({}, f, omitBy(to, isUndefined));
  const out = arr.slice();
  out.splice(fromIndex, 1, replacement); //{ ...to, label: f.label });
  return out;
};

/**
 * swap a point 'from', with point 'to'.
 * Finds the point in the array and updates it to equal the 'to' point.
 * @param {Point[]} points
 * @param {Point} from
 * @param {Point} to
 */

export const swapPoint = (points, from, to) => {
  return swap(points, from, to, pointIndex);
};
