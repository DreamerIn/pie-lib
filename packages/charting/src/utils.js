import { scaleBand, scalePoint } from '@vx/scale';
import { utils } from '@pie-lib/plot';

export const tickCount = utils.tickCount;
export const bounds = utils.bounds;
export const point = utils.point;

export const bandKey = (d, index) => `${index}-${d.label || '-'}`;

export const dataToXBand = (scaleX, data, width, type) => {
  switch (type) {
    case 'bar':
    case 'dotPlot':
    case 'linePlot':
      return scaleBand({
        rangeRound: [0, width],
        domain: data && data.map(bandKey),
        padding: 0.2
      });
    case 'histogram':
      return scaleBand({
        rangeRound: [0, width],
        domain: data && data.map(bandKey),
        padding: 0
      });
    case 'line':
      return scalePoint({
        domain: data && data.map(bandKey),
        rangeRound: [0, width]
      });
    default:
      return scaleBand({
        range: [0, width],
        domain: data && data.map(bandKey),
        padding: 0
      });
  }
};

export const getTickValues = (prop = {}) => {
  const tickValues = [];
  let tickVal = prop.min;

  while (tickVal <= prop.max) {
    tickValues.push(tickVal);
    tickVal = Math.round((tickVal + prop.step) * 100) / 100;
  }

  return tickValues;
};

export const getDomainAndRangeByChartType = (domain, range, chartType) => {
  const { min, max } = range || {};
  let { step } = range || {};
  const numberMax = parseFloat(max);

  if (!step) {
    if (numberMax < 20) {
      step = 1;
    } else if (numberMax >= 20 && numberMax < 100) {
      step = 5;
    } else {
      step = 10;
    }
  }

  switch (chartType) {
    // if chart is dot plot or line plot, we should ignore step and make sure that min & max are integer values
    case 'dotPlot':
    case 'linePlot':
      return {
        domain: {
          ...domain,
          step: 1,
          labelStep: 1,
          min: 0,
          max: 1
        },
        range: {
          ...range,
          min: parseInt(min, 10),
          max: parseInt(max, 10),
          step: 1
        }
      };
    default:
      return {
        domain: {
          ...domain,
          step: 1,
          labelStep: 1,
          min: 0,
          max: 1
        },
        range: {
          ...range,
          step
        }
      };
  }
};

export const getGridLinesAndAxisByChartType = (range, chartType) => {
  switch (chartType) {
    case 'line':
      return {
        verticalLines: undefined,
        horizontalLines: getTickValues(range),
        leftAxis: true
      };
    case 'dotPlot':
    case 'linePlot':
      return {
        verticalLines: [],
        horizontalLines: [],
        leftAxis: false
      };
    default:
      return {
        verticalLines: [],
        horizontalLines: getTickValues(range),
        leftAxis: true
      };
  }
};

export const getRotateAngle = barWidth => {
  if (barWidth < 30) {
    return 75;
  }

  if (barWidth < 40) {
    return 45;
  }

  if (barWidth < 60) {
    return 25;
  }

  return 0;
};

export const getTopPadding = barWidth => {
  if (barWidth < 30) {
    return 50;
  }

  if (barWidth < 40) {
    return 30;
  }

  if (barWidth < 60) {
    return 15;
  }

  return 0;
};
