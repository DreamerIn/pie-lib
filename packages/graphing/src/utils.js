import head from 'lodash/head';
import tail from 'lodash/tail';
import { utils } from '@pie-lib/plot';

export const tickCount = utils.tickCount;
export const bounds = utils.bounds;
export const point = utils.point;

// export { * from utils}
export const polygonToArea = points => {
  const h = head(points);
  const area = {
    left: h.x,
    top: h.y,
    bottom: h.y,
    right: h.x
  };
  return tail(points).reduce((a, p) => {
    a.left = Math.min(a.left, p.x);
    a.top = Math.max(a.top, p.y);
    a.bottom = Math.min(a.bottom, p.y);
    a.right = Math.max(a.right, p.x);
    return a;
  }, area);
};

export const lineToArea = (from, to) => {
  const left = Math.min(from.x, to.x);
  const top = Math.max(from.y, to.y);
  const bottom = Math.min(from.y, to.y);
  const right = Math.max(from.x, to.x);
  return { left, top, bottom, right };
};
