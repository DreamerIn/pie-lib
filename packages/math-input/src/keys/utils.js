export const mkSet = category => o => ({ ...o, category });

export const toArray = o => Object.keys(o).map(k => o[k]);

export const toKey = category => v => {
  if (typeof v === 'string') {
    return {
      name: v,
      label: v,
      category,
      write: v
    };
  } else {
    return {
      ...v,
      category
    };
  }
};
