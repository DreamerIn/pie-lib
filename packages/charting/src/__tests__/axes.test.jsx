import { shallow } from 'enzyme';
import React from 'react';
import ChartAxes from '../axes';
import { graphProps } from '../__tests__/utils';

describe('ChartAxes', () => {
  const wrapper = extras => {
    const defaults = {
      classes: {},
      className: 'className',
      graphProps: graphProps(),
      xBand: {
        bandwidth: () => {}
      }
    };
    const props = { ...defaults, ...extras };
    return shallow(<ChartAxes {...props} />);
  };
  describe('snapshot', () => {
    it('renders', () => expect(wrapper()).toMatchSnapshot());
  });
});
