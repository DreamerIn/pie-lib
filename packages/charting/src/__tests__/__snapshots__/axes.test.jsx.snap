// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartAxes snapshot renders 1`] = `
<RawChartAxes
  className="className"
  classes={
    Object {
      "axis": "RawChartAxes-axis-2",
      "axisLabel": "RawChartAxes-axisLabel-1",
      "axisLine": "RawChartAxes-axisLine-3",
      "tick": "RawChartAxes-tick-4",
    }
  }
  graphProps={
    Object {
      "domain": Object {
        "max": 1,
        "min": 0,
        "step": 1,
      },
      "range": Object {
        "max": 1,
        "min": 0,
        "step": 1,
      },
      "scale": Object {
        "x": [MockFunction],
        "y": [MockFunction],
      },
      "size": Object {
        "height": 400,
        "width": 400,
      },
      "snap": Object {
        "x": [MockFunction],
        "y": [MockFunction],
      },
    }
  }
  xBand={
    Object {
      "bandwidth": [Function],
    }
  }
/>
`;
